auth/src/main/java/org/aerogear/mobile/security/AsyncSecurityCheckExecutor.java

```diff
@@ -101,16 +105,37 @@ public AsyncSecurityCheckExecutor build() {
         final MetricsService metricsService = getMetricsService();
         final Map<String, Future<SecurityCheckResult>> res = new HashMap<>();
 
-        for (final SecurityCheck check : getChecks()) {
-            res.put(check.getName(), (executorService.submit(() -> {
-                final SecurityCheckResult result =  check.test(getContext());
-                if (metricsService != null) {
-                    metricsService.publish(new SecurityCheckResultMetric(result));
-                }
+        final Collection<SecurityCheck> checks = getChecks();
+        final int checkCount = checks.size();
+        for (final SecurityCheck check : checks) {
+            final Future<SecurityCheckResult> future = executorService.submit(() -> {
+                final SecurityCheckResult result = check.test(getContext());
                 return result;
-            })));
+            });
+            res.put(check.getName(), future);
+        }
+
+        if(metricsService != null) {
+            (new AppExecutors()).networkThread()
+                .submit(() -> publishMetricsAsync(checkCount));
         }
 
         return res;
     }
+
+    private void publishMetricsAsync(final int howMany) {
+        final CompletionService<SecurityCheckResult> cs = new ExecutorCompletionService<>(executorService);
+        final ArrayList<SecurityCheckResult> results = new ArrayList<>(howMany);
+        for (int i = 0; i < howMany; i++) {
+            try {
+                SecurityCheckResult result = cs.take().get();
```

Comment:
This seems to be the recommended way to consume these but I'm not sure if it's very solid.

@secondsun Is there a risk these could get mixed up with other tasks in the same `ExecutorService`?
It seems that by default it's the `AppExecutors#singleThreadService()` singleton and this method is set to be run in the `networkThread()`.

Reply:
What do you mean by "mixed up"?

Reply:
@secondsun maybe the race condition is only in my head.

IIUC the executorService is a queue so I thought that maybe if somewhere else in the main thread other kinds of jobs are queued into the `singleThreadService` the `cs.take()` call could end up picking a result that's meant for another async operation.

Reply:
Ahhh right.  So 1) The network thread is single threaded.  That isn't guaranteed, but it is right now.  2) If the order is important there are options for enforcing the order of the jobs, but I will have to refresh myself.  (Or we add a epic for weaving RxJava into the SDK implementation)

Reply:
@paolobueno With the suggested implementation, this problem does not exists